File: 0164-maximum-gap/0164-maximum-gap.rb
# @param {Integer[]} nums
# @return {Integer}
def maximum_gap(nums)
  g_min, g_max = nums.min, nums.max
  return 0 if nums.length < 2
  bucket_size = [1, (g_max-g_min)/(nums.length-1)].max
  buckets = Hash.new
  nums.each do |num|
    key = (num-g_min)/bucket_size
    if buckets.has_key?(key)
      buckets[key] = [[num, buckets[key][0]].min, [num, buckets[key][1]].max]
    else
      buckets[key] = [num, num]
    end
  end
  ans, pre_key = 0, -1
  buckets.keys.sort.each do |key|
    ans = [ans, buckets[key][0]-buckets[pre_key][1]].max if pre_key != -1
    pre_key = key
  end
  ans
end

File: 0164-maximum-gap/0164-maximum-gap.ts
function maximumGap(nums: number[]): number {
  const gMin: number = Math.min(...nums)
  const gMax: number = Math.max(...nums)
  if (nums.length < 2) return 0
  const bucketSize: number = Math.max(1, Math.floor((gMax - gMin) / (nums.length - 1)))
  const buckets = {}
  nums.forEach(num => {
    const key: number = Math.floor((num - gMin) / bucketSize)
    if (key in buckets) {
      buckets[key] = [Math.min(num, buckets[key][0]), Math.max(num, buckets[key][1])]
    } else {
      buckets[key] = [num, num]
    }
  })
  let ans = 0
  let preKey = -1
  const bucketKeys: number[] = Object.keys(buckets).map(d => +d)
  bucketKeys.sort((a, b) => a - b).forEach(key => {
    if (preKey !== -1) ans = Math.max(ans, buckets[key][0] - buckets[preKey][1])
    preKey = key
  })
  return ans
}